hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/Dispatcher.java

```diff
@@ -0,0 +1,272 @@
+import com.hedera.node.app.state.HederaState;
+import com.hedera.node.app.workflows.StoreCache;
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * A {@code Dispatcher} provides functionality to forward pre-check, pre-handle, and handle-requests
+ * to the appropriate handler
+ */
+public class Dispatcher {
+
+    private final Handlers handlers;
+
+    private final StoreCache storeCache;
+
+    /**
+     * Constructor of {@code Dispatcher}
+     *
+     * @param handlers a {@link Handlers} record with all available handlers
+     * @param storeCache a {@link StoreCache} that maintains stores for all active {@link
+     *     HederaState}s
+     * @throws NullPointerException if one of the parameters is {@code null}
+     */
+    public Dispatcher(@NonNull final Handlers handlers, @NonNull final StoreCache storeCache) {
+        this.handlers = requireNonNull(handlers);
+        this.storeCache = requireNonNull(storeCache);
+    }
+
+    /**
+     * Dispatch a pre-check request. It is forwarded to the correct handler, which takes care of the
+     * specific functionality.
+     *
+     * @param transactionBody the {@link TransactionBody} of the request
+     * @throws NullPointerException if {@code transactionBody} is {@code null}
+     * @throws PreCheckException if validation fails
+     */
+    public void preCheck(@NonNull final TransactionBody transactionBody) throws PreCheckException {
+        requireNonNull(transactionBody);
+
+        final var handler = getHandler(transactionBody);
+        handler.preCheck(transactionBody);
+    }
+
+    /**
+     * Dispatch a pre-handle request. It is forwarded to the correct handler, which takes care of
+     * the specific functionality. The payer is taken from the transaction.
+     *
+     * @param state the {@link HederaState} of this request
+     * @param transactionBody the {@link TransactionBody} of the request
+     * @throws NullPointerException if {@code transactionBody} is {@code null}
+     */
+    @NonNull
+    public TransactionMetadata preHandle(
+            @NonNull final HederaState state, @NonNull final TransactionBody transactionBody) {
+        requireNonNull(state);
+        requireNonNull(transactionBody);
+
+        final AccountID payer = transactionBody.getTransactionID().getAccountID();
+        return preHandle(state, transactionBody, payer);
+    }
+
+    /**
+     * Dispatch a pre-handle request. It is forwarded to the correct handler, which takes care of
+     * the specific functionality
+     *
+     * @param state the {@link HederaState} of this request
+     * @param transactionBody the {@link TransactionBody} of the request
+     * @param payer the {@link AccountID} of the payer
+     * @throws NullPointerException if one of the arguments is {@code null}
+     */
+    @NonNull
+    public TransactionMetadata preHandle(
```

Comment:
IMO dispatcher should just have dispatch method and `preHandle`, `preCheck` methods should be on `TransactionHandler`. Is there a reason we moved them to dispatcher ? 🤔